clever-tools/default.nix

```diff
@@ -2,21 +2,25 @@
 
 stdenv.mkDerivation rec {
   name = "clever-tools-${version}";
-  version = "0.9.3";
+  version = "0.10.1";
```

Comment:
Do you mean extracting the version as a parameter ? And then defaulting if not supplied ?

Reply:
For instance. I've seen another pattern, for instance in [the ansible expression](https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/admin/ansible/default.nix#L53-L64) which exposes various derivations, letting the main one defaulting to the last one.

Reply:
Okay I see, can you let me check tomorrow something that I have in mind ?
The pattern made me think on a possibly better solution, here she is: 
Imagine we have a derivation for the package version 0, which would be the first supported version (in our case, the first commit that allow us to build the package correctly).
Then, when committing a new version X, we just reference a parent commit, and possibly change the build process withouth impacting the version X-1 (defined in a previous commit).

Why this would be better ? In fact, we could have differents build processes for differents versions, which is supported with what I just said. In contrast to the ansible example where the build process **should** actually be the same for every versions (if it changes, they will have to create a second function for it ?).

I don't know if that is possible, and will try to make it work tomorrow :) !

Here is an illustration of what I think:

```nix
let mkCleverToolsDerivation = { ... };
in rec {
  clever-tools-0 = import (fetchFromGitHub
                   { owner  = "fretlink";
                     repo   = "clever-tools-nix";
                     rev    = "COMMIT_OF_DEFINED_VERSION";
                     sha256 = "...";
                   }.clever-tools);

  clever-tools-1 = import (fetchFromGitHub
                   { owner  = "fmenou";
                     repo   = "clever-tools-nix";
                     rev    = "COMMIT_OF_DEFINED_VERSION";
                     sha256 = "...";
                   }.clever-tools);

  # Notice that this variable `clever-tools` mean, for a commit,
  # the latest version available when the commit was made :)
  clever-tools = mkCleverToolsDerivation;
}
```

This is recursively fun !

Reply:
[Here is the draft](https://github.com/hussein-aitlahcen/clever-tools-nix/tree/exhaustive-versionning).
I've also updated the example in the branch:
```nix
{ pkgs ? import <nixpkgs> {} }: with pkgs;
let
  clever-tools = fetchFromGitHub {
                   owner  = "hussein-aitlahcen"; # assume fretlink
                   repo   = "clever-tools-nix";
                   rev    = "f4b7fca8a69144b6415f11c9218730d12ad9a542";
                   sha256 = "091k5srdl7f51c4xah2cvjq7wisml0bmzs8ychnk7g5hlc687s9m";
                 };
in (import clever-tools {}).v0_9_3
```
Here you can use `v0_9_3` or `latest` which would be `0.10.1`
You now have to set the script to `nix build -f example`, because the `default.nix` would install all versions.
Can you take a look at it and tell me what you think :).